objects/src/transaction/proven_tx.rs

```diff
@@ -159,6 +125,292 @@ impl ProvenTransaction {
+    pub fn add_input_notes<T>(mut self, notes: T) -> Self
+    where
+        T: IntoIterator<Item = Nullifier>,
+    {
+        self.input_notes.extend(notes);
+        self
+    }
+
+    /// Add notes produced by the transaction.
+    pub fn add_output_notes<T>(mut self, notes: T) -> Self
+    where
+        T: IntoIterator<Item = NoteEnvelope>,
+    {
+        self.output_notes.extend(notes);
+        self
+    }
+
+    /// Add produce notes details.
+    pub fn add_output_note_details<T>(mut self, notes: T) -> Self
+    where
+        T: IntoIterator<Item = (NoteId, Note)>,
+    {
+        self.output_note_details.extend(notes);
+        self
+    }
+
+    /// Set transaction's script root.
+    pub fn tx_script_root(mut self, tx_script_root: Digest) -> Self {
+        self.tx_script_root = Some(tx_script_root);
+        self
+    }
+
+    /// Set the transaction's proof.
+    pub fn proof(mut self, proof: ExecutionProof) -> Self {
+        self.proof = Some(proof);
+        self
+    }
+
+    /// Set reference block the transaction was executed against.
+    pub fn block_ref(mut self, block_ref: Digest) -> Self {
+        self.block_ref = Some(block_ref);
+        self
+    }
+
+    pub fn build(mut self) -> Result<ProvenTransaction, ProvenTransactionBuilderError> {
+        // Required fields
+        let account_id =
+            self.account_id.take().ok_or(ProvenTransactionBuilderError::MissingAccountId)?;
+        let initial_account_hash = self
+            .initial_account_hash
+            .take()
+            .ok_or(ProvenTransactionBuilderError::MissingInitialHash)?;
+        let final_account_hash = self
+            .final_account_hash
+            .take()
+            .ok_or(ProvenTransactionBuilderError::MissingFinalHash)?;
+        let block_ref = self
+            .block_ref
+            .take()
+            .ok_or(ProvenTransactionBuilderError::MissingBlockReference)?;
+        let proof =
+            self.proof.take().ok_or(ProvenTransactionBuilderError::MissingExecutionProof)?;
+
+        let output_note_details = self.output_note_details;
+        let known_output_ids = BTreeSet::from_iter(self.output_notes.iter().map(|n| n.note_id()));
+        let unknown_ids: Vec<_> = output_note_details
+            .keys()
+            .filter(|k| !known_output_ids.contains(k))
+            .cloned()
+            .collect();
+        if !unknown_ids.is_empty() {
+            return Err(ProvenTransactionBuilderError::NoteDetailsForUnknownNotes(unknown_ids));
+        }
+
+        let account_details = self.account_details.take();
+        let input_notes = InputNotes::new(self.input_notes)
+            .map_err(ProvenTransactionBuilderError::InputNotesError)?;
+        let output_notes = OutputNotes::new(self.output_notes)
+            .map_err(ProvenTransactionBuilderError::OutputNotesError)?;
+        let tx_script_root = self.tx_script_root;
+
+        if let Some(ref details) = account_details {
```

Comment:
We should be able to tell wither account details must be included based on the account ID. So, if account ID is for an on-chain account, but there are no account details, we should return an error.

Another error condition is when the account is new (i.e., the initial hash is all zeros) but the details are either missing or contain account delta.

To summarize:
- If an account is an on-chain account (based on ID), the transaction must contain account details.
  - If an account is a new account (based on init hash), the details must contain the `Account` struct and account ID / final hash must match.
  - Otherwise, the details must contain account delta.
- Otherwise, account details must be empty.

Btw, let's update the comment on line 39 above to document that default digest means that the account is new.

Reply:
Can't we simplify this?

- Make the account type explicit in the type system, `OnchainAccount`/`OffchainAccount`.
- Have only delta state changes, and force it to be every key that changed in a transaction

Reply:
updated the checks

Reply:
To extend on the above. A new account has an initial state, which are the trees with the leaves set to the empty word. And the initial state can be computed as the diff from that tree to the new state. Meaning we only need to have the delta changes.

Reply:
> Make the account type explicit in the type system, `OnchainAccount`/`OffchainAccount`

Were you thinking of using enums for this - e.g. something like:
```Rust
pub enum AccountId {
    OnChain(Felt),
    OffChain(Felt),
}
```
But I'm not sure how we can enforce that some bits in the underlying `Felt` specify which variant it is.

> Have only delta state changes, and force it to be every key that changed in a transaction

This could work, but we'd need to modify how `AccountDelta` works for this (iirc, we don't currently serialize the layout element of the storage as we assume that it doesn't changes once set). I would leave this for future PRs.

Reply:
Created https://github.com/0xPolygonMiden/miden-base/issues/524.